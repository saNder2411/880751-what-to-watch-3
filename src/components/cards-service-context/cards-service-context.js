import {createContext} from 'react';

const {
  Provider: CardsServiceProvider,
  Consumer: CardsServiceConsumer
} = createContext();

export {CardsServiceProvider, CardsServiceConsumer};
