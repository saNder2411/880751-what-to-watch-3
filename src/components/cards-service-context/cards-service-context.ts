import * as React from 'react';

const {
  Provider: CardsServiceProvider,
  Consumer: CardsServiceConsumer
} = React.createContext(void 0);

export {CardsServiceProvider, CardsServiceConsumer};
