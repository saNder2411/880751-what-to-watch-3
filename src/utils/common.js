const TimeShiftInMin = {
  YEAR: 525600,
  MONTH: 43800,
  WEEK: 10080,
  DAY: 1440,
  HOUR: 60,
  MINUTES: 1,
};

const Month = [`January`, `February`, `March`, `April`, `May`, `June`, `July`, `August`, `September`, `October`, `November`, `December`];

const MAX_RATING = 10;

export default class Common {
  static getRandomNumberFromPeriod(max, min = 0) {
    return min + Math.floor((max - min) * Math.random());
  }

  static getRandomDate(shiftInMin = TimeShiftInMin.YEAR * 20, sign = -1) {
    const currentDate = new Date();
    const diffValue = sign * this.getRandomNumberFromPeriod(shiftInMin);

    currentDate.setMinutes(currentDate.getMinutes() + diffValue);

    return +currentDate;
  }

  static getRandomRating() {
    const randomNumber = Common.getRandomNumberFromPeriod(MAX_RATING + 1, 1);
    const randomDecimal = Common.getRandomNumberFromPeriod(MAX_RATING);

    return randomNumber < MAX_RATING ? `${randomNumber}.${randomDecimal}` : `${randomNumber}`;
  }

  static calcCardLevel(ratingVal, [minVal, middleLowVal, middleVal, maxVal]) {
    const ratingToNum = +ratingVal;
    let level = `Bad`;

    if (ratingToNum >= minVal && ratingToNum <= middleLowVal) {
      level = `Normal`;
    } else if (ratingToNum >= middleLowVal && ratingToNum <= middleVal) {
      level = `Good`;
    } else if (ratingToNum >= middleVal && ratingToNum < maxVal) {
      level = `Very good`;
    } else if (ratingToNum >= maxVal) {
      level = `Awesome`;
    }

    return level;
  }

  static parseDateToStr(date) {
    const day = new Date(date).getDate();
    const month = new Date(date).getMonth();
    const year = new Date(date).getFullYear();

    return `${Month[month]} ${day}, ${year}`;
  }

  static dividedArrayInHalf(array) {
    const firstPartLength = ((array.length % 2) + array.length) / 2;

    return [array.slice(0, firstPartLength), array.slice(firstPartLength)];
  }

  static extend(a, b) {
    return Object.assign({}, a, b);
  }
}
