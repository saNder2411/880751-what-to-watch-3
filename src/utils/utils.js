const Month = [`January`, `February`, `March`, `April`, `May`, `June`, `July`, `August`, `September`, `October`, `November`, `December`];
const MINUTES_IN_HOUR = 60;

const calcCardLevel = (ratingVal, [minVal, middleLowVal, middleVal, maxVal]) => {
  const ratingToNum = +ratingVal;
  let level = `Bad`;

  if (ratingToNum >= minVal && ratingToNum <= middleLowVal) {
    level = `Normal`;
  } else if (ratingToNum >= middleLowVal && ratingToNum <= middleVal) {
    level = `Good`;
  } else if (ratingToNum >= middleVal && ratingToNum < maxVal) {
    level = `Very good`;
  } else if (ratingToNum >= maxVal) {
    level = `Awesome`;
  }

  return level;
};

const dividedArrayInHalf = (array) => {
  const firstPartLength = ((array.length % 2) + array.length) / 2;

  return [array.slice(0, firstPartLength), array.slice(firstPartLength)];
};

const parseDateToStr = (date) => {
  const day = new Date(date).getDate();
  const month = new Date(date).getMonth();
  const year = new Date(date).getFullYear();

  return `${Month[month]} ${day}, ${year}`;
};

const getTimeInHoursAndMinutes = (timeInMin) => {
  const runtime = {
    hours: `0`,
    minutes: `0`,
  };

  if (timeInMin < MINUTES_IN_HOUR) {
    runtime.minutes = (timeInMin < 10) ? `0${timeInMin}` : `${timeInMin}`;
    return runtime.minutes;
  }

  runtime.minutes = timeInMin % MINUTES_IN_HOUR;
  runtime.hours = (timeInMin - runtime.minutes) / MINUTES_IN_HOUR;

  return (runtime.hours) ? `${runtime.hours}h ${runtime.minutes}m` : `${runtime.minutes}m`;
};

const extend = (a, ...b) => Object.assign({}, a, ...b);

export {calcCardLevel, parseDateToStr, dividedArrayInHalf, getTimeInHoursAndMinutes, extend};
