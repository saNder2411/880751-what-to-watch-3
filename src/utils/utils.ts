const Month = [`January`, `February`, `March`, `April`, `May`, `June`, `July`, `August`, `September`, `October`, `November`, `December`];
const MINUTES_IN_HOUR = 60;

const calcCardLevel = (ratingVal: number, [minVal, middleLowVal, middleVal, maxVal]: Array<number>) => {
  let level = `Bad`;

  if (ratingVal >= minVal && ratingVal <= middleLowVal) {
    level = `Normal`;
  } else if (ratingVal >= middleLowVal && ratingVal <= middleVal) {
    level = `Good`;
  } else if (ratingVal >= middleVal && ratingVal < maxVal) {
    level = `Very good`;
  } else if (ratingVal >= maxVal) {
    level = `Awesome`;
  }

  return level;
};

const dividedArrayInHalf = (array) => {
  const firstPartLength = ((array.length % 2) + array.length) / 2;

  return [array.slice(0, firstPartLength), array.slice(firstPartLength)];
};

const parseDateToStr = (date) => {
  const day = new Date(date).getDate();
  const month = new Date(date).getMonth();
  const year = new Date(date).getFullYear();

  return `${Month[month]} ${day}, ${year}`;
};

const getTimeInHoursAndMinutes = (timeInMin: number) => {
  const runtime = {
    hours: `0`,
    minutes: `0`,
  };

  if (timeInMin < MINUTES_IN_HOUR) {
    runtime.minutes = (timeInMin < 10) ? `0${timeInMin}` : `${timeInMin}`;
    return runtime.minutes;
  }

  runtime.minutes = (timeInMin % MINUTES_IN_HOUR).toString();
  runtime.hours = ((timeInMin - (+runtime.minutes)) / MINUTES_IN_HOUR).toString();

  return (runtime.hours) ? `${runtime.hours}h ${runtime.minutes}m` : `${runtime.minutes}m`;
};

const extend = (a, ...b) => Object.assign({}, a, ...b);

const getAppRoute = (id = `:id`) => ({
  ROOT: `/`,
  CARDS: `/cards/${id}`,
  PLAYER: `/player/${id}`,
  LOGIN: `/login`,
  REVIEW: `/cards/${id}/review`,
  USER_LIST: `/userList`,
});

const noop = () => {};

export {calcCardLevel, parseDateToStr, dividedArrayInHalf, getTimeInHoursAndMinutes,
  extend, getAppRoute, noop};
