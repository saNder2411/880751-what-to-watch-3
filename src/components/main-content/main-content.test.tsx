import * as React from 'react';
import * as renderer from 'react-test-renderer';
import thunk from 'redux-thunk';
import createAPI from '../../api';
import {Provider} from 'react-redux';
import configureStore from 'redux-mock-store';
import {BrowserRouter, Switch, Route} from 'react-router-dom';
import CardsService from '../../services/cards-service';
import {CardsServiceProvider} from '../cards-service-context/cards-service-context';

import MainContent from './main-content';
import GenreList from '../genre-list/genre-list';
import PreviewCardList from '../preview-card-list/preview-card-list';
import ShowMoreButton from '../show-more-button/show-more-button';
import Footer from '../footer/footer';
import Logo from '../logo/logo';

import withPreviewCardListState from '../../hocs/with-preview-card-list-state/with-preview-card-list-state';
import withActiveItem from '../../hocs/with-active-item/with-active-item';
import compose from '../../hocs/compose/compose';
import {noop} from '../../utils/utils';


const API = createAPI(noop);
const cardsService = new CardsService(API);
const mockStore = configureStore([thunk]);

const mockPromoCardData = {
  id: 1,
  backgroundImage: `bg-the-grand-budapest-hotel`,
  posterImage: `the-grand-budapest-hotel-poster`,
  previewImage: `img/bohemian-rhapsody.jpg`,
  title: `Bohemian Rhapsody`,
  description: `In the 1930s, the Grand Budapest Hotel is a popular European ski resort, presided over by concierge Gustave H. (Ralph Fiennes). Zero, a junior lobby boy, becomes Gustave's friend and protege.`,
  rating: 9,
  scoresCount: 100,
  previewVideoSrc: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
  videoSrc: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
  director: `Steven Spielberg`,
  starring: [
    `Judi Dench`, `Robert De Niro`, `Leonardo DiCaprio`, `Morgan Freeman`, `Tom Hanks`,
  ],
  runtime: 98,
  isFavorite: true,
  genre: `Drama`,
  released: 1989,
};

const mockCardsData = [
  {
    id: 1,
    backgroundImage: `bg-the-grand-budapest-hotel`,
    posterImage: `the-grand-budapest-hotel-poster`,
    previewImage: `img/bohemian-rhapsody.jpg`,
    title: `Bohemian Rhapsody`,
    description: `In the 1930s, the Grand Budapest Hotel is a popular European ski resort, presided over by concierge Gustave H. (Ralph Fiennes). Zero, a junior lobby boy, becomes Gustave's friend and protege.`,
    rating: 9,
    scoresCount: 100,
    previewVideoSrc: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    videoSrc: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    director: `Steven Spielberg`,
    starring: [
      `Judi Dench`, `Robert De Niro`, `Leonardo DiCaprio`, `Morgan Freeman`, `Tom Hanks`,
    ],
    runtime: 98,
    isFavorite: true,
    genre: `Drama`,
    released: 1989,
  },
];
const store = mockStore({
  appState: {
    screen: `MAIN`,
    selectedCardId: 1,
  },
  user: {
    isAuthorized: false,
    userData: {},
    userDataLoading: true,
    userDataError: null,
  },
  userCardList: {
    userCardsData: [],
    userCardsLoading: false,
    userCardsError: null,
  },
  promoCard: {
    promoCardData: mockPromoCardData,
    promoLoading: false,
    promoError: null,
  },
  cardList: {
    cardsData: mockCardsData,
    cardsLoading: false,
    cardsError: null,
    updatedCardLoading: false,
    updatedCardError: null,
  },
  cardListState: {
    genre: `All genre`,
    showingCardsAmount: 8,
  },
  reviews: {
    reviewsData: [],
    reviewsLoading: false,
    isReviewAdded: false,
    reviewsError: null,
  }
});


const WrappedGenreList = withActiveItem(GenreList);
const WrappedPreviewCardList = compose(withActiveItem, withPreviewCardListState)(PreviewCardList);


it(`Should MainContent render correctly`, () => {
  const markup = renderer.create(
      <Provider store={store}>
        <CardsServiceProvider value={cardsService}>
          <BrowserRouter>
            <Switch>
              <Route
                path='/'
                exact
              >
                <MainContent>
                  <WrappedGenreList/>
                  <WrappedPreviewCardList />
                  <ShowMoreButton />
                  <Footer>
                    <Logo isFooterLogo/>
                  </Footer>
                </MainContent>
              </Route>
            </Switch>
          </BrowserRouter>
        </CardsServiceProvider>
      </Provider>,
      {
        createNodeMock: () => {
          return {};
        }
      }
  )
  .toJSON();

  expect(markup).toMatchSnapshot();
});
