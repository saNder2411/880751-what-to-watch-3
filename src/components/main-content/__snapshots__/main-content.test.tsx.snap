// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should MainContent render correctly 1`] = `
<div
  className="page-content"
>
  <section
    className="catalog"
  >
    <h2
      className="catalog__title visually-hidden"
    >
      Catalog
    </h2>
    <ul
      className="catalog__genres-list"
    >
      <li
        className="catalog__genres-item "
      >
        <a
          className="catalog__genres-link"
          href="#"
          onClick={[Function]}
        >
          All genres
        </a>
      </li>
      <li
        className="catalog__genres-item "
      >
        <a
          className="catalog__genres-link"
          href="#"
          onClick={[Function]}
        >
          Drama
        </a>
      </li>
    </ul>
    <div
      className="catalog__movies-list"
    />
  </section>
  <footer
    className="page-footer"
  >
    <div
      className="logo"
    >
      <a
        className="logo__link logo__link--light"
        href="/"
        onClick={[Function]}
      >
        <span
          className="logo__letter logo__letter--1"
        >
          W
        </span>
        <span
          className="logo__letter logo__letter--2"
        >
          T
        </span>
        <span
          className="logo__letter logo__letter--3"
        >
          W
        </span>
      </a>
    </div>
    <div
      className="copyright"
    >
      <p>
        Â© 2019 What to watch Ltd.
      </p>
    </div>
  </footer>
</div>
`;
